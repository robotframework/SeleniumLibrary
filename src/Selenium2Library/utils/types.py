# Originally based on  Robot Framework 3.0.2 robot.utils.robottypes
# Can be removed when library minimum required Robot Framework version is
# greater than 3.0.2. Then Robot Framework is_truthy should also support
# string NONE as Python False.
import sys


PY2 = sys.version_info[0] == 2

if PY2:
    def is_string(item):
        return isinstance(item, (str, unicode))
else:
    from robot.utils import is_string


def is_truthy(item):
    if is_string(item):
        return item.upper() not in ('FALSE', 'NO', '', 'NONE')
    return bool(item)


def is_falsy(item):
    return not is_truthy(item)
