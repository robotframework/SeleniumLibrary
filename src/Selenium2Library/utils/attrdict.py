class AttrDict(dict):

    def __getattr__(self, name):
        try:
            return self[name]
        except KeyError:
            raise AttributeError(name)

    def __setattr__(self, name, value):
        self[name] = value

    def __delattr__(self, name):
        try:
            del self[name]
        except KeyError:
            raise AttributeError(name)
