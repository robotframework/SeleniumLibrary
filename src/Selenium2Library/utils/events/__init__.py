from scope_event import ScopeStart, ScopeEnd

_registered_events = [ ScopeStart, ScopeEnd ]
_events = []

__all__ = [
    "on",
    "dispatch",
    "register_event"
]

def on(event_name, *args, **kwargs):
    for event in _registered_events:
        if event.name == event_name:
            _events.append(event(*args, **kwargs))
            return

def dispatch(event_name, *args, **kwargs):
    for event in _events:
        if event.name == event_name:
            event.trigger(*args, **kwargs)

def register_event(event):
    for registered_event in _registered_events:
        if event.name == registered_event.name:
            raise AttributeError("An event with the name " + event.name + " already exists.")
    _registered_events.append(event)
