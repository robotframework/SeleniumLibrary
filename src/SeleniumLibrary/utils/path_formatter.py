# Copyright 2008-2011 Nokia Networks
# Copyright 2011-2016 Ryan Tomac, Ed Manlove and contributors
# Copyright 2016-     Robot Framework Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


def _format_path(file_path, index):
    return file_path.format_map(_SafeFormatter(index=index))


class _SafeFormatter(dict):
    def __missing__(self, key):
        return f"{{{key}}}"
