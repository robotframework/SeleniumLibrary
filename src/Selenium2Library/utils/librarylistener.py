# Copyright 2008-2011 Nokia Networks
# Copyright 2011-2016 Ryan Tomac, Ed Manlove and contributors
# Copyright 2016-     Robot Framework Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from .events import dispatch


class LibraryListener(object):
    ROBOT_LISTENER_API_VERSION = 2

    def start_suite(self, name, attrs):
        dispatch('scope_start', attrs['longname'])

    def end_suite(self, name, attrs):
        dispatch('scope_end', attrs['longname'])

    def start_test(self, name, attrs):
        dispatch('scope_start', attrs['longname'])

    def end_test(self, name, attrs):
        dispatch('scope_end', attrs['longname'])
