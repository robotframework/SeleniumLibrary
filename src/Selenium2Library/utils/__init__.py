import os
import sys
from fnmatch import fnmatch
from .browsercache import BrowserCache

__all__ = [
    "get_child_packages_in",
    "get_module_names_under",
    "import_modules_under",
    "escape_xpath_value",
    "BrowserCache"
]

PY3 = sys.version_info[0] == 3
if PY3:
    unicode = str


# Public

def get_child_packages_in(root_dir, include_root_package_name=True, exclusions=None):
    packages = []
    root_package_str = os.path.basename(root_dir) + '.' if include_root_package_name else ""
    _discover_child_package_dirs(
        root_dir,
        _clean_exclusions(exclusions),
        lambda abs_path, relative_path, name: 
            packages.append(root_package_str + relative_path.replace(os.sep, '.')))
    return packages

def get_module_names_under(root_dir, include_root_package_name=True, exclusions=None, pattern=None):
    module_names = []
    root_package_str = os.path.basename(root_dir) + '.' if include_root_package_name else ""
    _discover_module_files_in(
        root_dir,
        _clean_exclusions(exclusions),
        pattern if pattern is not None else "*.*",
        lambda abs_path, relative_path, name: 
            module_names.append(root_package_str + os.path.splitext(relative_path)[0].replace(os.sep, '.')))
    return module_names

def import_modules_under(root_dir, include_root_package_name=True, exclusions=None, pattern=None):
    module_names = get_module_names_under(root_dir, include_root_package_name, exclusions, pattern)
    modules = [ __import__(module_name, globals(), locals(), ['*'], 0)
        for module_name in module_names ]
    return (module_names, modules)

def escape_xpath_value(value):
    value = unicode(value)
    if '"' in value and '\'' in value:
        parts_wo_apos = value.split('\'')
        return "concat('%s')" % "', \"'\", '".join(parts_wo_apos)
    if '\'' in value:
        return "\"%s\"" % value
    return "'%s'" % value

# Private

def _clean_exclusions(exclusions):
    if exclusions is None: exclusions = []
    if not isinstance(exclusions, list): exclusions = [ exclusions ]
    exclusions = [ os.sep + exclusion.lower().strip(os.sep) + os.sep
        for exclusion in exclusions ]
    return exclusions

def _discover_child_package_dirs(root_dir, exclusions, callback, relative_dir=None):
    relative_dir = relative_dir if relative_dir is not None else ''
    abs_dir = os.path.join(root_dir, relative_dir)
    for item in os.listdir(abs_dir):
        item_relative_path = os.path.join(relative_dir, item)
        item_abs_path = os.path.join(root_dir, item_relative_path)
        if os.path.isdir(item_abs_path):
            if os.path.exists(os.path.join(item_abs_path, "__init__.py")):
                exclusion_matches = [ exclusion for exclusion in exclusions 
                    if os.sep + item_relative_path.lower() + os.sep == exclusion ]
                if not exclusion_matches:
                    callback(item_abs_path, item_relative_path, item)
                    _discover_child_package_dirs(root_dir, exclusions, callback, item_relative_path)

def _discover_module_files_in(root_dir, exclusions, pattern, callback):
    def find_matching_files(relative_dir):
        abs_dir = os.path.join(root_dir, relative_dir)
        for item in os.listdir(abs_dir):
            item_relative_path = os.path.join(relative_dir, item)
            item_abs_path = os.path.join(root_dir, item_relative_path)
            if os.path.isfile(item_abs_path) and fnmatch(item, pattern):
                callback(item_abs_path, item_relative_path, item)

    find_matching_files('')

    _discover_child_package_dirs(
        root_dir,
        _clean_exclusions(exclusions),
        lambda abs_path, relative_path, name: find_matching_files(relative_path))
