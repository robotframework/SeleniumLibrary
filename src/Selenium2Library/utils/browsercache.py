from robot.utils import ConnectionCache


class BrowserCache(ConnectionCache):

    def __init__(self):
        ConnectionCache.__init__(self, no_current_msg='No current browser')
        self._closed = set()

    @property
    def browsers(self):
        return self._connections

    def get_open_browsers(self):
        open_browsers = []
        for browser in self._connections:
            if browser not in self._closed:
                open_browsers.append(browser)
        return open_browsers
    
    def close(self):
        if self.current:
            browser = self.current
            browser.quit()
            self.current = self._no_current
            self._closed.add(browser)

    def close_all(self):
        for browser in self._connections:
            if browser not in self._closed:
                browser.quit()
        self.empty_cache()
        return self.current
