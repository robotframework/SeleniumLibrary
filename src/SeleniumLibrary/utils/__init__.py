# Copyright 2008-2011 Nokia Networks
# Copyright 2011-2016 Ryan Tomac, Ed Manlove and contributors
# Copyright 2016-     Robot Framework Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from robot.utils import plural_or_not, secs_to_timestr, timestr_to_secs  # noqa

from .librarylistener import LibraryListener  # noqa
from .types import (
    is_falsy,
    is_noney,
    is_string,
    is_truthy,
    WINDOWS,
    _convert_timeout,
)  # noqa


def escape_xpath_value(value: str):
    value = str(value)
    if '"' in value and "'" in value:
        parts_wo_apos = value.split("'")
        escaped = "', \"'\", '".join(parts_wo_apos)
        return f"concat('{escaped}')"
    if "'" in value:
        return f'"{value}"'
    return f"'{value}'"
