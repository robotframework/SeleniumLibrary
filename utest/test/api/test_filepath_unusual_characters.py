import os

import pytest
from approvaltests.approvals import verify_all
from approvaltests.reporters.generic_diff_reporter_factory import (
    GenericDiffReporterFactory,
)
from robot.utils import WINDOWS

from SeleniumLibrary.utils.path_formatter import _format_path


@pytest.fixture(scope="module")
def reporter():
    path = os.path.dirname(__file__)
    reporter_json = os.path.abspath(
        os.path.join(path, "..", "approvals_reporters.json")
    )
    factory = GenericDiffReporterFactory()
    factory.load(reporter_json)
    return factory.get_first_working()


@pytest.mark.skipif(WINDOWS, reason="ApprovalTest do not support different line feeds")
def test_normal_file_path(reporter):
    results = []
    results.append(_format_path("/foo/file.log", 1))
    results.append(_format_path("/foo/file-{index}.log", 1))
    results.append(_format_path("/foo/file-{index}.log", "0001"))
    results.append(_format_path("/foo/file-{foo}.log", 1))
    results.append(_format_path("/{foo}/file-{index}.log", 1))
    results.append(_format_path("/foo/file-{index:03}.log", 1))
    results.append(_format_path("/foo/{index}-file-{index}.log", "1234"))
    results.append(_format_path("/foo/file-{in dex}.log", "1234"))
    results.append(_format_path("/foo/file-{in@dex}.log", "1234"))
    verify_all("Different file paths.", results, reporter=reporter)
