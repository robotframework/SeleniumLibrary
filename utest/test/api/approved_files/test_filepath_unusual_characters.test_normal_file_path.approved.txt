Different file paths.

0) /foo/file.log
1) /foo/file-1.log
2) /foo/file-0001.log
3) /foo/file-{foo}.log
4) /{foo}/file-1.log
5) /foo/file-001.log
6) /foo/1234-file-1234.log
7) /foo/file-{in dex}.log
8) /foo/file-{in@dex}.log
